aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp

```diff
@@ -260,5 +261,128 @@ Tensor NestedTensor_to_mask(const Tensor& nt, c10::optional<int64_t> mask_dim, c
+    int64_t head_dim{query.size(-1) / (num_heads * 3)};
+    int64_t Nnz_q{cumulative_sequence_length_q[-1].item<int64_t>()};
+
+    // The input/output of flash attention is (Nnz, nheads, headdim) we need to
+    // create a nested size with the last dim = nheads*headdim If q,k,v are all
+    // equal than the current nt last dim is of size nheads*headdim*3 after
+    // projection
+    auto atten_size = get_nested_size_tensor(query);
+    auto div3 = atten_size.index({at::indexing::Slice(), -1}).div(3);
+    atten_size.index({at::indexing::Slice(), -1}) = div3;
+
+    auto qkv_buffer_reshaped =
+        get_buffer(query).view({Nnz_q, 3, num_heads, head_dim});
+
+    // If we are passing in query, key, value all the same tensors than we have
+    // packed them into one tensor and need to slice for flash attention
+    Tensor atten_buffer = at::_flash_scaled_dot_product_attention(
+        qkv_buffer_reshaped.index({at::indexing::Slice(), 0}),
+        qkv_buffer_reshaped.index({at::indexing::Slice(), 1}),
+        qkv_buffer_reshaped.index({at::indexing::Slice(), 2}),
+        cumulative_sequence_length_q,
+        cumulative_sequence_length_q,
+        max_seqlen_batch_q,
+        max_seqlen_batch_q,
+        dropout_p,
+        causal);
+    // Output of flash_attention is a regular tensor lets wrap it back up to
+    // form a nested tensor
+    return wrap_buffer(atten_buffer.view(-1), atten_size);
+  }
+
+  // Query, Key, and Value are not all the same tensor and therefore need to
+  // calculate K meta data
+
+  // The nested tensors will be of shape {Batch_size x ragged_seq_len x num_heads * head_dim }
+  auto cumulative_and_max_k = cumulative_and_max_seq_len(key);
+  Tensor cumulative_sequence_length_k = std::get<0>(cumulative_and_max_k);
+  int64_t max_seqlen_batch_k = std::get<1>(cumulative_and_max_k);
+
+  // K and V have to have the same Nnz, should probably torch_check before now assume in order to not iterate over v
+  int64_t Nnz_q{cumulative_sequence_length_q[-1].item<int64_t>()};
+  int64_t Nnz_kv{cumulative_sequence_length_k[-1].item<int64_t>()};
+
+  int64_t head_dim{query.size(-1) / num_heads};
+
+  auto query_buffer_reshaped = get_buffer(query).view({Nnz_q, num_heads, head_dim});
+  auto key_buffer_reshaped = get_buffer(key).view({Nnz_kv, num_heads, head_dim});
+  auto value_buffer_reshaped = get_buffer(value).view({Nnz_kv, num_heads, head_dim});
+
+  Tensor atten_buffer = at::_flash_scaled_dot_product_attention(
+        query_buffer_reshaped,
+        key_buffer_reshaped,
+        value_buffer_reshaped,
+        cumulative_sequence_length_q,
+        cumulative_sequence_length_k,
+        max_seqlen_batch_q,
+        max_seqlen_batch_k,
+        dropout_p,
+        causal);
+  // Output of flash_attention is a regular tensor lets wrap it back up to
+  // form a nested tensor, the size of which should match the value tensor
+  return wrap_buffer(atten_buffer.view(-1), get_nested_size_tensor(value));
```

Comment:
Is this right? Fairly sure size should match the query tensor.

Reply:
just updated